File: 0278-first-bad-version/0278-first-bad-version.go
/** 
 * Forward declaration of isBadVersion API.
 * @param   version   your guess about first bad version
 * @return 	 	      true if current version is bad 
 *			          false if current version is good
 * func isBadVersion(version int) bool;
 */
func lowerBound(n int) int {
    start := 1
    end := n
    for;start<end;{
        mid := int((start + end)/2)
        if isBadVersion(mid) {
            end = mid
        } else {
            start = mid
        }
        if end - start == 1 {
            break
        }
    }
    return end
}
func firstBadVersion(n int) int {
    if isBadVersion(1) {
        return 1
    }
    return lowerBound(n)
}

File: 0278-first-bad-version/0278-first-bad-version.py
# The isBadVersion API is already defined for you.
# def isBadVersion(version: int) -> bool:

class Solution:
    def minBadSearcher(self,n):
        start = 1
        end = n
        while start < end:
            mid = (start + end) // 2
            if isBadVersion(mid):
                end = mid
            else:
                start = mid
            if end - start == 1:
                return end

    def firstBadVersion(self, n: int) -> int:
        if isBadVersion(1): return 1
        return self.minBadSearcher(n)
        